runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/GroupByKeyTranslator.java

```diff
@@ -189,13 +194,44 @@ public ByteBuffer apply(WindowedValue<KV<K, V>> wv) {
       org.joda.time.Instant t1 = kv1.getKey();
       org.joda.time.Instant t2 = kv2.getKey();
 
-      WindowedValue<KV<K, Iterable<V>>> wv1 = kv1.getValue();
-      WindowedValue<KV<K, Iterable<V>>> wv2 = kv2.getValue();
+      final WindowedValue<KV<K, Iterable<V>>> wv1 = kv1.getValue();
+      final WindowedValue<KV<K, Iterable<V>>> wv2 = kv2.getValue();
+
+      Collection<? extends BoundedWindow> mergedWindows = wv1.getWindows();
+      if (!isNonMerging) {
+        mergedWindows =
+            mergeWindows((Collection<IntervalWindow>) wv1.getWindows(),
+                (Collection<IntervalWindow>) wv2.getWindows());
+      }
 
-      return KV.of(outputTimeFn.combine(t1, t2),
+      org.joda.time.Instant timestamp = outputTimeFn.combine(t1, t2);
+      return KV.of(timestamp,
           WindowedValue.of(KV.of(wv1.getValue().getKey(),
-              Iterables.concat(wv1.getValue().getValue(), wv2.getValue().getValue())),
-              wv1.getTimestamp(), wv1.getWindows(), wv1.getPane()));
+              Iterables.concat(wv1.getValue().getValue(), wv2.getValue().getValue())), timestamp,
+              mergedWindows, wv1.getPane()));
+    }
+
+    private Collection<? extends BoundedWindow> mergeWindows(
+        Collection<IntervalWindow> windows1, Collection<IntervalWindow> windows2) {
+      ArrayList<IntervalWindow> lists = new ArrayList<>();
+      lists.addAll(windows1);
+      lists.addAll(windows2);
+      Collections.sort(lists);
+
+      ArrayList<IntervalWindow> merges = new ArrayList<>();
```

Comment:
This implements merging for `Sessions`, but it should really be generic, using the user's `WindowFn` to merge.

If you want to support only sessions, that is fine (via capability matrix) but I think you also need some checks to make sure the input transform only uses sessions.